receiver/receivercreator/observerhandler.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package receivercreator
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/jwangsadinata/go-multimap"
+	"github.com/open-telemetry/opentelemetry-collector/component"
+	"github.com/open-telemetry/opentelemetry-collector/component/componenterror"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/observer"
+)
+
+var _ observer.Notify = (*observerHandler)(nil)
+
+// observerHandler manages endpoint change notifications.
+type observerHandler struct {
+	sync.Mutex
+	logger *zap.Logger
+	// receiverTemplates maps receiver template full name to a receiverTemplate value.
+	receiverTemplates map[string]receiverTemplate
+	// receiversByEndpointID is a map of endpoint IDs to a receiver instance.
+	receiversByEndpointID multimap.MultiMap
```

Comment:
Why is this a multimap? Can we have more than one receiver per endpoint ID? Is it because we can have different receiver types associated with the same ID? Should this be a map of maps: `map[endpointID]map[receiverFullName]receiver` in that case or we don't care about having the nested map?

Reply:
> Why is this a multimap? Can we have more than one receiver per endpoint ID?

Right, multiple receivers can be started for a single endpoint. The multimap is basically just a wrapper around a map of maps. I could do my own light wrapper instead but it's a lightweight library with no other dependencies.

Reply:
OK, up to you. I thought it might be nicer to have stronger compoile-time typing from map of maps. Your choice.

Reply:
Yeah the lack of type checking is unfortunate. Maybe one day Go will bless us with generics. :) I'll try doing custom wrapper, doesn't need all possible multimap functions so should be limited.